### PR TITLE
Correct behavior with storyshots

### PR DESCRIPTION
**Pull request by nicbillen**
At the moment there is an issue with storyshots. I get the following exception (line numbers can differ because it was from my local version with some extra log messages):

|         return kind + "/" + story;
      76 |     }
    > 77 |     throw new Error();
         |           ^
      78 | }
      79 | function waitForDelayTime(time) {
      80 |     if (time === void 0) { time = 0; }

      at url2StoryKey (../../itcb/storycap/packages/storycap/lib/client/trigger-screenshot.js:77:11)
      at url2StoryKey (../../itcb/storycap/packages/storycap/lib/client/trigger-screenshot.js:126:16)
      at Object.stock [as triggerScreenshot] (../../itcb/storycap/packages/storycap/lib/client/trigger-screenshot.js:200:5)
      at triggerScreenshot (../../itcb/storycap/packages/storycap/lib/client/with-screenshot.js:16:34)
      at ../../itcb/storycap/packages/storycap/node_modules/@storybook/addons/dist/make-decorator.js:37:14
      at ../../itcb/storycap/packages/storycap/node_modules/@storybook/addons/dist/make-decorator.js:61:46
      at node_modules/@storybook/client-api/node_modules/@storybook/addons/dist/hooks.js:238:21
      at node_modules/@storybook/client-api/dist/client_api.js:129:14
      at node_modules/@storybook/client-api/dist/client_api.js:130:16
      at withSubscriptionTracking (node_modules/@storybook/client-api/dist/client_api.js:158:16)
      at node_modules/@storybook/client-api/node_modules/@storybook/addons/dist/hooks.js:238:21
      at node_modules/@storybook/client-api/dist/client_api.js:129:14
      at node_modules/@storybook/client-api/node_modules/@storybook/addons/dist/hooks.js:266:20
      at Object.storyFn (node_modules/@storybook/client-api/dist/story_store.js:363:30)
      at getRenderedTree (node_modules/@storybook/addon-storyshots/dist/frameworks/react/renderTree.js:22:30)
      at node_modules/@storybook/addon-storyshots/dist/test-bodies.js:21:18
      at Object.<anonymous> (node_modules/@storybook/addon-storyshots/dist/api/snapshotsTestsTemplate.js:44:33)

For fixing this problem, I now read story and kind out of the context.

### CODE REVIEW

**Comment by Quramy** [comment]
Hi @nicbillen . Thanks for this PR.

I'm checkout your branch and run e2e test , and found the following errors:

```
$ ./e2e.sh examples/v5-managed-react

...
warn Capture timeout exceeded in 5000 msec. Retry to screenshot this story after this sequence. ImageButton default
error Screenshot timeout exceeded. 'capture' function is not triggerd in 5000 ms. Target story: Welcome/to Storybook
ScreenshotTimeoutError: Screenshot timeout exceeded. 'capture' function is not triggerd in 5000 ms. Target story: Welcome/to Storybook
    at Timeout.setTimeout [as _onTimeout] (/Users/yosuke/workspace/reg-viz/storycap/packages/storycap/lib/node/capturing-browser.js:144:28)
    at ontimeout (timers.js:436:11)
    at tryOnTimeout (timers.js:300:5)
    at listOnTimeout (timers.js:263:5)
    at Timer.processTimers (timers.js:223:10)
error Command failed with exit code 1.
```

I think your change does not work with Storybook v5.x 🤔  Can is the `context` object enabled in Storybook v5 ?  

Quramy

**Comment by nicbillen** [comment]
Hmm. That's a problem. We use "@storybook/addon-storyshots": "^5.2.8", "@storybook/react": "^5.2.0" and some addons more. I think I have to read a little bit more of the storybook code to give you an answer. Does storyshots work in your case (with the master branch?). I get these errors, if I call "npm run test" with our configuration.

**Comment by nicbillen** [comment]
I think the storyshots plugin sets thes context. But there are no url parameters available at this time (so the url2StoryKey will fail). What is your idea to solve the problem? Introducing an extra check for the context object or is there another way to get story and key? I have to test the e2e tests on another day.

**Comment by Quramy** [comment]
Hmmm, check whether `context` exist in url2StoryKey function ?

File: packages/storycap/src/client/trigger-screenshot.ts [hunk 1]
```diff
@@ -75,8 +75,15 @@ function consumeOptions(win: StorycapWindow, storyKey: string): ScreenshotOption
   return result;
 }
 
-function stock(opt: ScreenshotOptions = {}) {
-  const storyKey = url2StoryKey(location.href);
+function stock(opt: ScreenshotOptions = {}, context: any) {
+  let storyKey: string | undefined = undefined;
+
+  if (!!context.kind && !!context.story) {
```
Comment (Quramy) [inline packages/storycap/src/client/trigger-screenshot.ts]: Might be the `context` undefined ? 

```ts
if (context && !!context.kind && !!context.story) {
```

Comment (nicbillen) [reply packages/storycap/src/client/trigger-screenshot.ts]: Ok. I have added this extra check. Does the e2e test work now?

**Comment by Quramy** [comment]
Hmmmm. It seems that In Storybook v5, the `context` object always exist... I think we need other way to detect "whether it's running in storyshots"

**Comment by nicbillen** [comment]
But are there a context.kind an context.story? I can not explain, why it runs into a timout. In case of an undefined context, there should be the old behaviour and context, context.story and context.kind should be undefined. Which values are in your context object?

**Comment by Quramy** [comment]
I think `withScreenshotDecorator` must do nothing (must only delegate to original getStory fn) if Storybook runs by storyshots because storyshots runs in not web browser but node-env jest.

```ts
const withScreenshotDecorator =
  makeDecorator &&
  makeDecorator({
    name: 'withScreenshot',
    parameterName: 'screenshot',
    skipIfNoParametersOrOptions: false,
    allowDeprecatedUsage: true,
    wrapper: (getStory, context, { parameters, options }) => {
      if (isRunningInStoryshots() ) return getStory(context);
      const screenshotOptions = parameters || options;
      triggerScreenshot(screenshotOptions, context);
      return getStory(context);
    },
  });
```

**Comment by Quramy** [comment]
> context.kind an context.story

I think `kind` and `story` parameters does not exist in SB v5.  SB v5 `id` parameter instead of them. see  https://github.com/reg-viz/storycap/blob/master/packages/storycap/src/client/trigger-screenshot.ts#L32-L37 . 

( This change has hurt me when I did v5 support ... 😇 )

**Comment by nicbillen** [comment]
Yes, that can be the case. But is the id also in the context object? And in that case I would expect, that if (context && !!context.kind && !!context.story) is false and it switches to the old behaviour. I will do the e2e test soon by myself and post my results.

**Comment by nicbillen** [comment]
Now I have tested it a little bit with the examples. Indeed, it was the id. Now I take it from the context object.

**Comment by Quramy** [comment]
@nicbillen Thanks! I've published your change as Storycap v2.3.3 😄

**Pull request merged**